cpp/ql/lib/semmle/code/cpp/commons/StringConcatenation.qll

```diff
@@ -0,0 +1,96 @@
+/**
+ * A library for detecting general string concatenations.
+ */
+
+import cpp
+import semmle.code.cpp.models.implementations.Strcat
+import semmle.code.cpp.models.interfaces.FormattingFunction
+private import semmle.code.cpp.dataflow.new.DataFlow
+
+class StringConcatenation extends Call {
+  StringConcatenation() {
+    // sprintf-like functions, i.e., concat through formatting
+    this instanceof FormattingFunctionCall
+    or
+    this.getTarget() instanceof StrcatFunction
+    or
+    this.getTarget() instanceof StrlcatFunction
+    or
+    // operator+ concat
+    exists(Call call, Operator op |
+      call.getTarget() = op and
+      op.hasQualifiedName(["std", "bsl"], "operator+") and
+      op.getType().(UserType).hasQualifiedName(["std", "bsl"], "basic_string") and
+      this = call
+    )
+    or
+    // string stream concat (operator<<)
+    this.getTarget().hasQualifiedName(["std", "bsl"], "operator<<")
```

Comment:
My only worry is making sure the type is correct. Is it sufficient to say the type of the argument is a char or basic_string to cover all bases or is that too general? My general thought on this was simply to call the `getAnOperand` predicate which would ensure we find an operand of the right type, but I don't think I can do that from the char pred.

Reply:
What do you think of a solution of instead of worrying about the type of the operands we focus on where the << or + operator are defined? 
``
basic_ostream<char, char_traits<char>>
basic_string<char, char_traits<char>, allocator<char>>
```

I've found that the + and << operator are instantiations of a template with the above stripped type name (<< and + respectively). What if we say we are explicitly looking for concats of these template types? I'm playing with this currently but I'm not sure I can cleanly get the template base name out of the type yet (there has to be a way), but if forced to, I can just get the name and look that it starts with `basic_ostream` and `basic_string`. 

What do you think of this approach?

Reply:
Great. I'll get that change in shortly.

Reply:
@geoffw0  and @MathiasVP I've updated the code to check for ostream. I think this is ready for merge otherwise.